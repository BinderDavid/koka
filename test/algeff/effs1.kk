public module effs1

effect amb {
  flip() : bool
}
 
function xor() : amb bool {
  val p = flip() 
  val q = flip()
  (p||q) && not(p&&q)
}

val amb = handler {
  return x -> [x]
  flip()   -> resume(False) + resume(True)
}

function main() {
  amb(xor).showList(show).println
}

