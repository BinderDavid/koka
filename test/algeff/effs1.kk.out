[False,True,True,False]
.Op-flip : .op-flip
.Ops-flip : (flip : .op-flip) -> .ops-amb
.tag-amb : string
.tag-op-flip: string
amb : forall<a,e> (action : () -> <amb|e> a) -> e list<a>
flip : () -> amb bool
main : () -> console ()
xor : () -> amb bool